core/src/env/engine/on_chain/ext.rs

```diff
@@ -16,58 +16,133 @@
 //!
 //! Refer to substrate SRML contract module for more documentation.
 
-use crate::env::{
-    EnvError,
-    Result,
-};
 use ink_primitives::Key;
+use crate::env::ReturnFlags;
 
-/// Returned by the host environment if a contract call trapped.
-const TRAP_RETURN_CODE: u32 = 0x0100;
+macro_rules! define_error_codes {
+    (
+        $(
+            $( #[$attr:meta] )*
+            $name:ident = $discr:literal,
+        )*
+    ) => {
+        /// Either success or any error that can be returned from a runtime API call.
+        #[repr(u32)]
+        pub enum ReturnCode {
+            /// API call successful.
+            Success = 0,
+            $(
+                $( #[$attr] )*
+                $name = $discr,
+            )*
+            /// Returns if an unknown error was received from the host module.
+            UnknownError,
+        }
+
+        /// Every error that can be returned to a contract when it calls any of the host functions.
+        #[repr(u32)]
+        pub enum Error {
+            $(
+                $( #[$attr] )*
+                $name = $discr,
+            )*
+            /// Returns if an unknown error was received from the host module.
+            UnknownError,
+        }
+
+        impl From<ReturnCode> for Result {
+            #[inline]
+            fn from(return_code: ReturnCode) -> Self {
+                match return_code {
+                    ReturnCode::Success => Ok(()),
+                    $(
+                        ReturnCode::$name => Err(Error::$name),
+                    )*
+                    ReturnCode::UnknownError => Err(Error::UnknownError),
```

Comment:
How would that work? Seal does not produce this variant. What you need is a generated:
```
impl From<u32> for Result
```

where `_ => Error::UnknownError`.